File: 958-CheckCompletenessofaBinaryTree/958-CheckCompletenessofaBinaryTree.java
/*

958. Check Completeness of a Binary Tree
Medium

Given the root of a binary tree, determine if it is a complete binary tree.

In a complete binary tree, every level, except possibly the last, is completely filled, and all nodes in the last level are as far left as possible. It can have between 1 and 2h nodes inclusive at the last level h.

Example 1:

Input: root = [1,2,3,4,5,6]
Output: true
Explanation: Every level before the last is full (ie. levels with node-values {1} and {2, 3}), and all nodes in the last level ({4, 5, 6}) are as far left as possible.

Example 2:

Input: root = [1,2,3,4,5,null,7]
Output: false
Explanation: The node with value 7 isn't as far left as possible.

Constraints:

    The number of nodes in the tree is in the range [1, 100].
    1 <= Node.val <= 1000

*/

class Solution {
    public boolean isCompleteTree(TreeNode root) {
        if(root == null) return true;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean nullNode = false;
        while(!q.isEmpty()){
            int len = q.size();
            for(int i = 0; i < len; i++){
                TreeNode curr = q.poll();
                if(curr == null) nullNode = true;
                else{
                    if(nullNode) return false;
                    q.add(curr.left);
                    q.add(curr.right);
                }
            }
        }
        return true;
    }
}

File: 958-CheckCompletenessofaBinaryTree/958-CheckCompletenessofaBinaryTree.py
"""

958. Check Completeness of a Binary Tree

Given the root of a binary tree, determine if it is a complete binary tree.

In a complete binary tree, every level, except possibly the last, is completely filled, and all nodes in the last level are as far left as possible. It can have between 1 and 2h nodes inclusive at the last level h.

Example 1:

Input: root = [1,2,3,4,5,6]
Output: true
Explanation: Every level before the last is full (ie. levels with node-values {1} and {2, 3}), and all nodes in the last level ({4, 5, 6}) are as far left as possible.

Example 2:

Input: root = [1,2,3,4,5,null,7]
Output: false
Explanation: The node with value 7 isn't as far left as possible.

Constraints:

    The number of nodes in the tree is in the range [1, 100].
    1 <= Node.val <= 1000

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
        q = deque()
        if not root: return True
        q.append(root)
        nullNode = False
        while q:
            lenn = len(q)
            for _ in range(lenn):
                curr = q.popleft()
                if not curr: nullNode = True
                else:
                    if nullNode: return False
                    q.append(curr.left)
                    q.append(curr.right)
        return True
